content/topics/unit-testing/tdd-horrors/_index.md

```diff
@@ -0,0 +1,214 @@
+---
+title: TDD Horrors
+ready: true
+---
+
+# Common TDD pitfalls.
+These are all the most common TDD problems recruits tend to come across, read and make sure you didn't make any of these or any mistakes at all.
+
+- **Write tests**. Recruits in the past have done the mistake of not writing tests for TDD assignments, make sure you avoid this weird mistake. always write tests for your TDD projects.
+- No code in your TDD file, This is a repetition of the first point, but it's worth it. write tests people.
+- Files naming convention, name your files properly and according to what's inside them. E.g. if you have a file called add.js that only contains a multiply function then something is wrong. Do this:
+<!-- examples -->
+
+**Python**:
+
+`add.py`
+```py
+def add(***):
+  # some operations
+```
+**Javascript**
+
+`add.js`
+```js
+function add(***){
+  // operations
+}
+```
+
+- Defining a test case(testing function) inside a function. That will pass, even when it's supposed to fail.
+
+---
+**Python example:**
+
+```py
+def testOne():
+
+  def testSomething():
+    assert add("1,23,5") == 25
+
+  def testSomething():
+    assert add("1,23,5") == 29
+
+```
+instead you could put your tests into classes as methods, like so:
+
+```py
+class TestOne():
+
+  def testSomething():
+    assert add("1,23,5") == 25
+
+  def testSomething():
+    assert add("1,23,5") == 29
+
+```
+notice **def** keyword in the first cell is replaced with the **class** keyword in the second cell, this difference makes a major difference. The first case would **pass** as a single test, even though it contains two test cases and one of them must fail, in contrary the second case would run both tests, the first one would fail and the second one would pass.
+*Note:*
+Although it's not necessary to put stuff inside classes. One nice thing about the `pytest` module is that it removes the need for boilerplate code. so the following code would also work:
+```py
+  from module import add
+
+  # tests start here.
+  def test_add_two_values():
+    assert add(1,2) == 3
+
+  def test_add_multiple_values():
+    assert add(1,2,3,4) == 10
+```
+
+---
+
+- Test classes/describe function should focus on one functionality and one functionality only.
+
+---
+**Python example:**
+
+Do not do this:
+```py
+  class TestFunctions():
+
+    def test_addition(self):
+      assert add("1,2,3,4") == 10
+
+    def test_multiply(self):
+      assert multiply("1,2,3") == 9
+
+```
+Do this instead:
+```py
+# test add function.
+class TestAdd():
+  """docstring for TestAdd"""
+
+  def test_add_two_values(self):
+    assert add("1,2") == 3
+
+  def test_add_multiple_values(self):
+    assert add("1,2,3,4,5,6") == 21
+    assert add("2,3,4,5") == 14
+
+# test multiply function.
+class TestMultiply():
```

Comment:
rather delete this

Reply:
flat is better than nested. 
```
project dir -> test dir -> test.py -> class TestSomething
                                                  -> class TestSomethingElse
```

It's true that the above would work, but if the number of test functions is significant enough to actually make these classes useful for catagorisation of the tests, then the following would be a better structure:

```
project dir -> test dir -> test_something.py 
                                 -> test_something_else.py
```
Other pros to this approach:
- doesn't teach people bad oop habits. Classes aren't just to make your code look hardcore
- makes navigation from code easier
- tests are easier to reason about

Most of the time when people start a big project with the class approach they regret it. The only exception is if the `unittest` module is in use.